feat(identity): Principal directory reference (#5327)
Kong/developer.konghq.com: app/kong-identity/principals-and-directories.md

---
title: "Principals and directories"
content_type: reference
layout: reference

> permalink: /kong-identity/principals/

Author: ```suggestion
permalink: /identity/principals/
```

products:
  - konnect

works_on:
  - konnect

breadcrumbs:

>   - /kong-identity/

Author: ```suggestion
  - /identity/
```

---
title: "Principals and directories"
content_type: reference
layout: reference
permalink: /identity/principals/
products:
  - identity
  - konnect

works_on:
  - konnect

breadcrumbs:
  - /identity/

description: "Learn how {{site.identity}} principals and directories represent authenticating entities and centralize identity, credentials, and metadata across {{site.base_gateway}}, {{site.event_gateway_short}}, and {{site.dev_portal}}."

related_resources:
  - text: "{{site.identity}}"
    url: /kong-identity/
  - text: "Centrally-managed Consumers"

>     url: /gateway/entities/consumer/#centrally-managed-consumers

Reviewer: Let's drop this. We'll be pushing most CMS customers to Principals going forward, and starting to hide this.

  - text: "{{site.konnect_short_name}} control plane resource limits"
    url: /gateway/control-plane-resource-limits/
---

{{site.identity}} uses principals and directories to unify how Kong products represent the entities they authenticate. 

> * **Principal:** Represents an entity that authenticates to a Kong product. 

Reviewer: s/a Kong Product/a Kong Gateway/

(so as to not confuse with Konnect Users or Developer Portal Developers which are different concepts entirely)

  The entity can be a human, a workload acting on behalf of a human, or a workload acting on behalf of itself (machine authentication).
* **Directory:** Regional collection of principals. 

>   Each {{site.konnect_short_name}} organization gets one directory per region.

Reviewer: s/gets one directory per region/can provision up to one directory per region by default/

Principals can be used for:

> * {{site.base_gateway}} in {{site.konnect_short_name}}: Can both authenticate clients against a principal and look up principal metadata after authentication

Reviewer: I'd suggest something like:

Principals can be used for:
* In Kong Gateway in Konnect: Can be used to authenticate an entity for API traffic (using API Keys or Basic Auth usernames and passwords) _or_ can be used to provide metadata about an entity for OAuth-authenticated traffic.
* In Event Gateway: Can be used metadata about an authenticated entity
* In Developer Portal: Can be used to adjust API Gateway behavior based on the authenticating Developer Portal Application

* {{site.event_gateway_short}}: Can only look up principal metadata after authentication has happened at the broker
* {{site.dev_portal}}: Applications can be mapped to a principal to apply Consumer-scoped plugins to application traffic

The following table describes when you'd want to use principals:

<!--vale off-->
{% table %}
columns:
  - title: Use case
    key: usecase
  - title: Description
    key: description
rows:

>   - usecase: "Authenticate clients at {{site.base_gateway}}"

Reviewer: 'clients' is an OAuth term that we also use elsewhere in Kong Identity. Maybe use "authenticate API traffic" instead?

    description: "Terminate authentication at {{site.base_gateway}} using credentials stored on a principal, such as a username and password or an API key."
  - usecase: "Hydrate metadata after external authentication"
    description: "After a third-party IdP verifies an OAuth token, look up the matching principal in a directory to retrieve metadata for policy decisions."
  - usecase: "Share identity across Kong products"
    description: "Define a principal once and use it for {{site.base_gateway}}, {{site.event_gateway_short}}, and {{site.dev_portal}} without manually synchronizing identities."
  - usecase: "Drive gateway behavior with metadata"
    description: "Scope rate limits, ACLs, and conditional plugin execution by attributes that the principal carries, such as business unit or tier."
{% endtable %}
<!--vale on-->

## Principal configuration

Every principal carries some combination of credentials, identities, and metadata. 

### Credentials

> Credentials are secrets a principal presents at authentication time. The gateway checks them directly against the directory.

Reviewer: ...against the directory and authenticate the principal.

* **Scale beyond Consumer limits:** Users with large populations of authenticating entities can exceed the [Consumer control plane limits](/gateway/control-plane-resource-limits/). Consumers are loaded into data plane memory while principals are loaded on demand when they authenticate.
* **Share identity across gateways without manual sync:** When the same entity needs to authenticate to multiple gateways, Consumers must be replicated. A principal lives in one directory and can be referenced from any gateway in the same region.
* **Unify identity across Kong products:** Consumers and applications each represent the same logical entity in different ways, which makes it hard to reason about authentication, policy, and observability together. Principals replace that with one model.
* **Attach metadata to identity:** Consumers don't support metadata, which forces custom plugins or header-injection workarounds when you want per-entity rate limits, conditional plugin behavior, or richer logs. Principals support metadata natively.

If you have existing Consumers you want to migrate to principals, see [HOW TO PLACEHOLDER](/).

### Metadata

> Metadata is a set of key/value pairs attached to a principal. Values can be strings, integers, floats, booleans, datetimes, or lists of any of those types.

Reviewer: Could use a richer description here, eg.

Metadata is a set of key/value pairs attached to a principal that are made available in the request context of a Kong API or Event Gateway after that principal has been authenticated, or looked up, in the request context.

> You can use metadata for policy decisions. 

Reviewer: This is a little open-ended. Maybe:

_You can use metadata to configure how a Gateway behaves when that Principal is authenticated to it._

For example:

> * A rate limiting plugin can filter by `principal.metadata.tier`.

Reviewer: This is a great example, but won't actually be true until 3.16 :(

* An ACL plugin can allow access only when `principal.metadata.business_unit == "payments"`.

> * A logging plugin can include `principal.metadata.region` in every log line.

Reviewer: This is a great example, but won't actually be true until 3.16 :(

> ## Apply plugins to principals

Reviewer: I actually expected this heading to talk about how to configure the api key, basic auth or ODIC plugins. While we obviously don't want a full how-to should we have a section that briefly mentions those?

This is important too but I'd suggest a more specific heading, eg. "Controlling API Gateway Plugin Execution based on Principals"

Author: Yep! I'll add those in an examples section.

> ## Apply plugins to principals

Reviewer: Where do we talk about configuring event GW virtual cluster authentication or policies?

Plugins can read principal attributes from the request context after authentication. 
The only way to scope plugins to a principal is by using conditional plugin execution on the principal metadata.

For example, a rate limiting plugin scoped to gold-tier principals would use a condition like this:

```yaml
condition: principal.metadata.tier == "gold"
```

The following request context fields are available to plugins:

* `principal.id`: The UUID of the principal
* `principal.metadata.*`: Any metadata key on the principal
* `principal.identities.*`: Any identity on the principal

> ## Principal limitations

Reviewer: Should this go at the end?

    description: "Scope conditional plugin execution by attributes that the principal carries, such as business unit or tier."
{% endtable %}
<!--vale on-->

## Principal configuration

Every principal carries some combination of credentials, identities, and metadata. 

### Credentials

Credentials are secrets a principal presents at authentication time. 
You need an API Gateway configured in {{site.konnect_short_name}} to use credentials because the gateway checks them directly against the directory and authenticates the principal.

{{site.identity}} supports the following credential types:

* **Username and password**: A principal can have up to two passwords per username to support rotation. Principals in the same directory can't share a username.
* **API key**: Each API key must be unique within a directory. 
  API keys can be system-generated by {{site.identity}} or imported with a custom secret value.

A single principal can have multiple credentials, including multiple credentials of the same type. 
To authenticate using a credential, a principal must exist in the target directory with a matching credential set.

### Identities

> Identities are references to how a principal is known in other systems. 

Reviewer 2: Avoiding passive voice for a smoother reading:

```suggestion
Identities are references to how other systems identify a principal. 
```

> They are used for lookup after authentication has already happened somewhere else (like a third-party IdP), or to link a principal to an entity in another Kong product. 

Reviewer 2: ```suggestion
{{site.identity}} uses them either:

- For lookup after authentication has already happened somewhere else (like a third-party IdP).
- To link a principal to an entity in another Kong product. 
```

> Identities are created on the principal as a separate resource after the principal is created.

Reviewer 2: ```suggestion
You can add identities to the principal as a separate resource after you've created a principal.
```

rows:
  - type: "OIDC"
    api: "`oidc`"
    description: "References a principal authenticated by an external IdP, such as Okta or Cognito. Identified by the token's issuer URL and a configurable claim (claim name and value). Most commonly the claim is `sub`, but any claim in the token can be used for lookup."
  - type: "Auth server client"
    api: "`auth_server_client`"
    description: "References a [{{site.identity}} auth server](/identity/) client in the same region. Identified by the auth server ID and client ID."
  - type: "Consumer"
    api: "`control_plane_consumer`"
    description: "Links the principal to a Consumer in a specific {{site.base_gateway}} control plane. Use this identity to apply existing Consumer-scoped plugins to a principal's traffic."
  - type: "Custom"
    api: "`custom`"
    description: "A user-defined key and value, used to look up a principal by an identifier from an external system. For example, a `sasl_username` key with value `john` lets {{site.event_gateway_short}} match a SASL-authenticated Kafka client to a principal."
{% endtable %}

A single principal can have multiple identities, including multiple of the same type.

{:.info}

> > **Lookup caching and time-to-live (TTL):** Principal lookups are cached, so additional lookups don't require a connection to {{site.identity}} until the cache is evicted. For {{site.base_gateway}}, the eviction rate is determined by the smallest of the following:

Reviewer 2: ```suggestion
> **Lookup caching and time-to-live (TTL):** {{site.identity}} caches principal lookups, so additional lookups don't require a connection to {{site.identity}} until it evicts the cache. For {{site.base_gateway}}, the eviction rate depends on the smallest of the following:
```

> * The TTL set on the directory (10 minutes by default)
> * The TTL set on a credential
> * The caching configuration on {{site.base_gateway}}. Cached principals are in the general cache pool which can be configured by setting `db_cache_ttl`.

### Principal entity mapping

Principals centralize the concept of an authenticating entity across Kong products. 
Each product has its own representation of who is authenticating: {{site.base_gateway}} has Consumers and {{site.dev_portal}} has applications:

* **Consumers**: Attach a `control_plane_consumer` [identity](#identities) to map a principal to a Consumer in a specific {{site.base_gateway}} control plane. When an authentication plugin authenticates the principal, the mapped Consumer loads into the request context just as if the Consumer had been authenticated directly. This allows existing Consumer-scoped plugins to function while you migrate to principals.

> * **Applications**: A {{site.dev_portal}} application can be mapped to a {{site.base_gateway}} Consumer through a principal, creating a 1:1:1 relationship between the application, the principal, and the Consumer. This is how you apply Consumer-scoped plugins (including ACE and KAA) to traffic from a {{site.dev_portal}} application: configure the plugin on the mapped Consumer, and it runs for any request authenticated as the application. Consumer-dimension analytics also include the application's activity once the mapping is in place. A Portal Admin maps an existing application to an existing Consumer; {{site.identity}} creates or updates the principal of type `application` behind the scenes.

Reviewer 2: ```suggestion
* **Applications**: You can map a {{site.dev_portal}} application to a {{site.base_gateway}} Consumer through a principal, creating a 1:1:1 relationship between the application, the principal, and the Consumer. This is how you apply Consumer-scoped plugins (including ACE and KAA) to traffic from a {{site.dev_portal}} application: configure the plugin on the mapped Consumer, and it runs for any request authenticated as the application. Consumer-dimension analytics also include the application's activity once the mapping is in place. A Portal Admin maps an existing application to an existing Consumer; {{site.identity}} creates or updates the principal of type `application` behind the scenes.
```

#### When to use principals instead of Consumers

{{site.base_gateway}} Consumers have limitations that principals can address:

* **Scale beyond Consumer limits:** Users with large populations of authenticating entities can exceed the [Consumer control plane limits](/gateway/control-plane-resource-limits/). Consumers are loaded into data plane memory while principals are loaded on demand when they authenticate.
* **Share identity across gateways without manual sync:** When the same entity needs to authenticate to multiple gateways, Consumers must be replicated. A principal lives in one directory and can be referenced from any gateway in the same region.
* **Unify identity across Kong products:** Consumers and applications each represent the same logical entity in different ways, which makes it hard to reason about authentication, policy, and observability together. Principals replace that with one model.
* **Attach metadata to identity:** Consumers don't support metadata, which forces custom plugins or header-injection workarounds when you want per-entity rate limits, conditional plugin behavior, or richer logs. Principals support metadata natively.

### Metadata

> Metadata is a set of key/value pairs attached to a principal that are made available in the request context of a {{site.base_gateway}} or {{site.event_gateway_short}} after that principal has been authenticated, or looked up, in the request context.

Reviewer 2: ```suggestion
Metadata is a set of key-value pairs associated with a principal. {site.identity}} authenticates or looks up a principal, then makes the metadata available in the request context of a {{site.base_gateway}} or {{site.event_gateway_short}}.
```

The block references the directory to look up and the identity key to match the authenticated value against. 
For example, a `custom` identity on `sasl_username` matches the SASL username sent by the Kafka client. 
When a match is found, the principal's metadata is attached to `context.auth.principal.metadata.*` for the lifetime of the connection.

A [Modify Headers](/event-gateway/policies/modify-headers/) policy that only runs for principals on the operators team would use a condition like this:

```yaml
condition: context.auth.principal.metadata.team == "operators"
```
{% comment %}
For a complete walkthrough, see [Enrich Kafka SASL PLAIN connections with Kong Identity principal metadata](/event-gateway/kong-identity-metadata-integration/).
{% endcomment %}

## Create a directory

{% navtabs "create-directory" %}
{% navtab "API" %}

Send a `POST` request to the `/v2/directories` endpoint:

<!--vale off-->
{% konnect_api_request %}
url: /v2/directories
status_code: 201
method: POST
body:
  name: example-directory
  description: Example directory for principals
  allow_all_control_planes: true
{% endkonnect_api_request %}
<!--vale on-->

{% endnavtab %}
{% navtab "UI" %}

> A default directory is automatically created when you [create your first principal](#configure-a-principal) in the UI.

Reviewer 2: ```suggestion
{{site.identity}} automatically creates a default directory when you [create your first principal](#configure-a-principal) in the UI.
```

  1. Set when the credential expires:
      * To keep the credential from expiring, select **Never**.
      * To set an expiration, select **Set date and time**, then select a time range.
{% endnavtab %}
{% navtab "OAuth client" %}
  1. From the **Authentication type** dropdown menu, select "OAuth client".
  1. Select an auth server option:
      * To let Kong manage the OAuth provider and handle token issuance and validation, select **Use Kong Identity auth server** and do the following:
        1. From the **Authorization server** dropdown menu, select your authorization server.
        1. From the **Client** dropdown menu, select your client.
      * To connect to your existing OAuth provider, select **Use external auth server** and do the following:
        1. In the **Issuer** field, enter the issuer URL of the external identity provider.
        1. In the **Claim name** field, enter the OIDC claim used to identify the principal. Use sub for the subject identifier.
        1. In the **Claim value** field, enter the expected value of the selected OIDC claim. For sub, this is the subject identifier.
{% endnavtab %}

> {% navtab "Consumer" %}

Reviewer 2: ```suggestion
{% navtab "Consumer" %}
{:.info}
   > If you weren't using Consumer before using Principals, you won't see the Consumer option available, since Principals are meant to replace Consumers in {{site.konnect_short_name}}.
```

Author: ```suggestion
{% navtab "Consumer" %}
{:.info}
   > If you weren't using Consumers before using principals, you won't see the Consumer option available, since principals are meant to replace Consumers in {{site.konnect_short_name}}.
```

  1. From the **Authentication type** dropdown menu, select "Consumer linking".
  1. From the **Control plane** dropdown menu, select the control plane where the Consumer exists.
  1. From the **Consumer** dropdown menu, select the Consumer that represents this principal.
{% endnavtab %}
{% navtab "External identifier" %}
  1. From the **Authentication type** dropdown menu, select "External identifier".
  1. In the **Name** field, enter unique key that defines the identifier type.
  1. In the **Value** field, enter the exact name of the value in your external system.
{% endnavtab %}
{% endnavtabs %}
1. (Optional) To add more than one authentication type, click **Add and create another** and repeat the configuration for the additional type.
1. Click **Add**.
1. Click **Next**.
1. Add metadata to the principal by entering a key/value pair.
1. Click **Create**.

{% endnavtab %}
{% endnavtabs %}

## Principal limitations

The following are default limits for principals and directories:

> * 1 directory per region

Reviewer 2: ```suggestion
* 1 directory per region.
```

The metadata lives on the principal, not on the credential, so you can change it without editing the broker.

To fetch principal metadata on a [virtual cluster](/event-gateway/entities/virtual-cluster/), add a `fetch_kong_identity_principal` block to the cluster's authentication configuration. 
The block references the directory to look up and the identity key to match the authenticated value against. 
For example, a `custom` identity on `sasl_username` matches the SASL username sent by the Kafka client. 
When a match is found, the principal's metadata is attached to `context.auth.principal.metadata.*` for the lifetime of the connection.

A [Modify Headers](/event-gateway/policies/modify-headers/) policy that only runs for principals on the operators team would use a condition like this:

```yaml
condition: context.auth.principal.metadata.team == "operators"
```
{% comment %}
For a complete walkthrough, see [Enrich Kafka SASL PLAIN connections with Kong Identity principal metadata](/event-gateway/kong-identity-metadata-integration/).
{% endcomment %}

## Create a directory

{% navtabs "create-directory" %}

> {% navtab "API" %}

Reviewer 3: A user can provision a directory in the UI as well, so I do think it's worth noting that here. If a user has not created the directory in the API, in the UI, when the user creates their first principal, a default directory is created on their behalf.

Author: @jdemartini81 Happy to add UI instructions for the directory! Could you point me to where in the UI (or Figma mockups) where a user can configure the directory? I couldn't find it in the UI or the Unification Figma mockups.

Reviewer 3: It automatically is created when a user creates their first principal. My suggestion is to describe this behavior. For instance, if you have not yet created your directory via the API, to create it via the UI, simply create your first principal. We will create a "default" directory on your behalf. (You have a better flair for words than me.)

---
title: "Principals and directories"
content_type: reference
layout: reference
permalink: /identity/principals/
products:
  - identity
  - konnect

works_on:
  - konnect

breadcrumbs:
  - /identity/

description: "Learn how {{site.identity}} principals and directories represent authenticating entities and centralize identity, credentials, and metadata across {{site.base_gateway}}, {{site.event_gateway_short}}, and {{site.dev_portal}}."

related_resources:
  - text: "{{site.identity}}"
    url: /kong-identity/
  - text: "{{site.konnect_short_name}} control plane resource limits"

>     url: /gateway/control-plane-resource-limits/

Reviewer: Should we link to the Konnect API spec for Kong Identity too?

---

{{site.identity}} uses principals and directories to unify how Kong products represent the entities they authenticate. 

> * **Principal:** Represents an external client or workload (not a human {{site.konnect_short_name}} user or {{site.dev_portal}} developer) that authenticates to a {{site.base_gateway}}.

Reviewer: This is a little misleading. A principal _can_ reprresent a human - but one authenticating to a customer's Gateway. But it is a different entity from a Dev Portal Developer or a Konnect User.

* **Directory:** Regional collection of principals. 
  Each {{site.konnect_short_name}} organization can provision up to one directory per region by default.

Principals can be used for:

> * {{site.base_gateway}} in {{site.konnect_short_name}}: Can be used to authenticate an entity for API traffic (using API keys or basic auth usernames and passwords) or can be used to provide metadata about an entity for OAuth-authenticated traffic.

Reviewer: > entity for OAuth-authenticated traffic

We are going to do this for other authentication schemes too (eg. mTLS) in future releases so it might be worth re-wording this to  something like "can be used to provide metadata about an entity already authenticated through another mechanism (such as OAuth)"

> * {{site.event_gateway_short}}: Can use metadata about an authenticated entity

Reviewer: Suggest something like "can be used to provide metadata about an entity already authenticated through another mechanism (such as SASL/SCRAM or SASL/OAUTHBEARER)

* {{site.dev_portal}}: Can be used to adjust API gateway behavior based on the authenticating {{site.dev_portal}} application

The following table describes when you'd want to use principals:

<!--vale off-->
{% table %}
columns:
  - title: Use case
    key: usecase
  - title: Description
    key: description
rows:
  - usecase: "Authenticate API traffic for {{site.base_gateway}}"
    description: "Terminate authentication at {{site.base_gateway}} using credentials stored on a principal, such as a username and password or an API key."
  - usecase: "Hydrate metadata after external authentication"
    description: "After a third-party IdP verifies an OAuth token, look up the matching principal in a directory to retrieve metadata for policy decisions."
  - usecase: "Share identity across Kong products"
    description: "Define a principal once and use it for {{site.base_gateway}}, {{site.event_gateway_short}}, and {{site.dev_portal}} without manually synchronizing identities."
  - usecase: "Drive gateway behavior with metadata"

>     description: "Scope rate limits, ACLs, and conditional plugin execution by attributes that the principal carries, such as business unit or tier."

Reviewer: ACLs and rate limits aren't actually coming until 3.16. Conditional plugin execution is there though.

{% endtable %}
<!--vale on-->

## Principal configuration

Every principal carries some combination of credentials, identities, and metadata. 

> ### Credentials

Reviewer: We should add a note somewhere here that to use Credentials will require a Gateway to send these secrets to Konnect.

Credentials are secrets a principal presents at authentication time. The gateway checks them directly against the directory and authenticates the principal.

{{site.identity}} supports the following credential types:

> * **Username and password**: A principal can have up to two passwords per username to support rotation.

Reviewer: Also worth noting, no two principals in the same directory can have the same username

#### When to use principals instead of Consumers

{{site.base_gateway}} Consumers have limitations that principals can address:

* **Scale beyond Consumer limits:** Users with large populations of authenticating entities can exceed the [Consumer control plane limits](/gateway/control-plane-resource-limits/). Consumers are loaded into data plane memory while principals are loaded on demand when they authenticate.
* **Share identity across gateways without manual sync:** When the same entity needs to authenticate to multiple gateways, Consumers must be replicated. A principal lives in one directory and can be referenced from any gateway in the same region.
* **Unify identity across Kong products:** Consumers and applications each represent the same logical entity in different ways, which makes it hard to reason about authentication, policy, and observability together. Principals replace that with one model.
* **Attach metadata to identity:** Consumers don't support metadata, which forces custom plugins or header-injection workarounds when you want per-entity rate limits, conditional plugin behavior, or richer logs. Principals support metadata natively.

> If you have existing Consumers you want to migrate to principals, see [HOW TO PLACEHOLDER](/).

Reviewer: I think we decided this was going to be a fast follow :)

### Metadata

Metadata is a set of key/value pairs attached to a principal that are made available in the request context of a {{site.base_gateway}} or {{site.event_gateway_short}} after that principal has been authenticated, or looked up, in the request context.

You can use metadata to configure how a gateway behaves when that principal is authenticated to it.

> For example, an ACL plugin can allow access only when `principal.metadata.business_unit == "payments"`.

Reviewer: ACL plugin does not _yet_ explicitly support Principals. However you can add a conditional expression to the Request Termination plugin.

## Controlling {{site.base_gateway}} plugin execution based on principals

Plugins can read principal attributes from the request context after authentication. 

> The only way to scope plugins to a principal is by using conditional plugin execution on the principal metadata.

Reviewer: Worth linking to the conditional plugin execution docs.

For example, a rate limiting plugin scoped to gold-tier principals would use a condition like this:

```yaml
condition: principal.metadata.tier == "gold"
```

The following request context fields are available to plugins:

* `principal.id`: The UUID of the principal
* `principal.metadata.*`: Any metadata key on the principal

> * `principal.identities.*`: Any identity on the principal

Reviewer: Can we double check this is true with @jijiechen ? If so, what is the schema of the different identity objects?

        1. In the **Claim value** field, enter the expected value of the selected OIDC claim. For sub, this is the subject identifier.
{% endnavtab %}
{% navtab "Consumer" %}
  1. From the **Authentication type** dropdown menu, select "Consumer linking".
  1. From the **Control plane** dropdown menu, select the control plane where the Consumer exists.
  1. From the **Consumer** dropdown menu, select the Consumer that represents this principal.
{% endnavtab %}
{% navtab "External identifier" %}
  1. From the **Authentication type** dropdown menu, select "External identifier".
  1. In the **Name** field, enter unique key that defines the identifier type.
  1. In the **Value** field, enter the exact name of the value in your external system.
{% endnavtab %}
{% endnavtabs %}
1. (Optional) To add more than one authentication type, click **Add and create another** and repeat the configuration for the additional type.
1. Click **Add**.
1. Click **Next**.
1. Add metadata to the principal by entering a key/value pair.
1. Click **Create**.

{% endnavtab %}
{% endnavtabs %}

## Principal limitations

> The following are default limits for principals and directories:

Reviewer: Also, mention a limit of 100 authentication or principal lookup requests per second per Directory, not counting cached requests.

> * 1 directory per region

Reviewer: Add "Contact Kong Support if you need more than one Directory."

> * Soft limit of 100,000 principals per directory. Contact Kong to raise this limit.

Reviewer: Remove references to "soft limits". ie. 

"100,000 principals per directory. Contact Kong Support if you need higher limits."

Use this when {{site.identity}} is issuing the OAuth tokens and you want to map the issuing client to a principal.

Send a `POST` request to the `/v2/directories/{directoryId}/principals/{principalId}/identities` endpoint:

<!--vale off-->
{% konnect_api_request %}
url: /v2/directories/$DIRECTORY_ID/principals/$PRINCIPAL_ID/identities
status_code: 201
method: POST
body:
  type: auth_server_client
  auth_server_id: $AUTH_SERVER_ID
  client_id: $CLIENT_ID
{% endkonnect_api_request %}
<!--vale on-->

### Link a principal to a Consumer

Map a principal to a Consumer in a specific {{site.base_gateway}} control plane. 
Use this when you want existing Consumer-scoped plugins to run for traffic authenticated as the principal.

Send a `POST` request to the `/v2/directories/{directoryId}/principals/{principalId}/identities` endpoint:

<!--vale off-->
{% konnect_api_request %}
url: /v2/directories/$DIRECTORY_ID/principals/$PRINCIPAL_ID/identities
status_code: 201
method: POST
body:
  type: control_plane_consumer
  control_plane_id: $CONTROL_PLANE_ID
  consumer_id: $CONSUMER_ID
{% endkonnect_api_request %}
<!--vale on-->

### Link a principal to a Consumer Group

> placeholder, need array config

Reviewer: @jijiechen can help here.
